modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteClusterSnapshotCheckTest.java

```diff
@@ -584,6 +634,673 @@ public void testClusterSnapshotCheckMultipleTimes() throws Exception {
+                );
+
+                if (secondDelay == null) {
+                    discoSpi(grid(0)).unblock();
+
+                    fut.get(getTestTimeout());
+
+                    return;
+                }
+
+                discoSpi(grid(0)).blockNextAndRelease(msg -> msg instanceof FullMessage
+                    && ((FullMessage<?>)msg).type() == secondDelay.ordinal());
+
+                discoSpi(grid(0)).waitBlocked(getTestTimeout());
+
+                if (step2preparation != null)
+                    step2preparation.run();
+
+                assertThrowsAnyCause(
+                    log,
+                    fut2::get,
+                    IllegalStateException.class,
+                    "Validation of snapshot '" + SNAPSHOT_NAME + "' has already started"
+                );
+
+                discoSpi(grid(0)).unblock();
+
+                fut.get(getTestTimeout());
+            }
+            else {
+                if (waitForBothFirstDelays) {
+                    discoSpi(grid(0)).waitBlockedSize(2, getTestTimeout());
+
+                    if (secondDelay != null) {
+                        discoSpi(grid(0)).blockNextAndRelease(msg -> msg instanceof FullMessage
+                            && ((FullMessage<?>)msg).type() == secondDelay.ordinal());
+
+                        discoSpi(grid(0)).waitBlockedSize(2, getTestTimeout());
+                    }
+                }
+                else {
+                    if (secondDelay != null) {
+                        discoSpi(grid(0)).blockNextAndRelease(msg -> msg instanceof FullMessage
+                            && ((FullMessage<?>)msg).type() == secondDelay.ordinal());
+
+                        discoSpi(grid(0)).waitBlocked(getTestTimeout());
+                    }
+                    else
+                        fut2.get();
+                }
+
+                discoSpi(grid(0)).unblock();
+
+                fut2.get();
+
+                fut.get();
+            }
+        }
+        finally {
+            discoSpi(grid(0)).unblock();
+
+            if (cleaner != null)
+                cleaner.run();
+
+            awaitPartitionMapExchange();
+        }
+    }
+
+    /**  */
+    private void doTestNodeStopsDuringSnapshotChecking(int originatorIdx, int nodeToStopIdx, Set<Integer> stopped) throws Exception {
+        int grids = G.allGrids().size();
+
+        ClusterNode leaving = grid(nodeToStopIdx).cluster().localNode();
+
+        boolean requiredLeft = originatorIdx == nodeToStopIdx || grid(nodeToStopIdx).cluster().currentBaselineTopology()
+            .stream()
+            .anyMatch(bl -> bl.consistentId().equals(leaving.consistentId()));
+
+        int coordIdx = -1;
+
+        for (int i = 0; i < grids; ++i) {
+            if (stopped.contains(i))
+                continue;
+
+            final IgniteEx g = grid(i);
+
+            // Wait for all nodes complete checking.
+            assertTrue(
```

Comment:
This looks wierd. We should not wait before the operation. Why? Let's wait after, where we check. We can use a dedicated method like `waitClusterFinishedSnapshotCheck(Collection<UUID> excludedNodes)`

Reply:
> We should not wait before the operation. Why?

If don't wait it will lead to "Validation of snapshot has already started", due to `contexts` collections is cleaned after `clusterOpFut` completes.

> Let's wait after, where we check

Don't understand exactly your idea.